### PR TITLE
http payload

### PR DESCRIPTION
**Pull request by nozik**

### CODE REVIEW

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy [hunk 1]
```diff
@@ -132,6 +135,16 @@ class SpringBootBasedTest extends HttpServerTest<ConfigurableApplicationContext>
...
     assertTraces(1) {
+      def httpSpan = testRunner().getExportedSpans()[1]
+      def requestBody = (String) httpSpan.attributes.asMap().find {
+        e -> e.key.key == "http.request.body"
+      }.value
+      assert requestBody.startsWith("username=test&password=")
+      def responseBody = (String) httpSpan.attributes.asMap().find {
+        e -> e.key.key == "http.response.body"
+      }.value
+      assert requestBody.startsWith("")
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy]: why is it empty?

Comment (LironKS) [reply instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy]: this is the type of the request

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy [hunk 2]
```diff
@@ -30,6 +30,8 @@ import static io.opentelemetry.instrumentation.testing.junit.http.ServerEndpoint
 
 class SpringBootBasedTest extends HttpServerTest<ConfigurableApplicationContext> implements AgentTestTrait {
 
+  public static final String TEST_REQUEST_HEADER = "X-Test-Request";
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy]: delete

Comment (LironKS) [reply instrumentation/spring/spring-webmvc-3.1/javaagent/src/test/groovy/test/boot/SpringBootBasedTest.groovy]: done

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java [hunk 3]
```diff
@@ -72,22 +84,61 @@ public void init(FilterConfig filterConfig) {}
   public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
       throws ServletException, IOException {
 
+    ContentCachingResponseWrapper wrapper = new ContentCachingResponseWrapper((HttpServletResponse) response);
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java]: Don't we want to protect this cast? Is it 100% safe?

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java [hunk 4]
```diff
@@ -72,22 +84,61 @@ public void init(FilterConfig filterConfig) {}
...
       return;
     }
 
     try {
-      filterChain.doFilter(request, response);
+      filterChain.doFilter(request, wrapper);
     } finally {
       if (handlerMappings != null) {
         Context context = Context.current();
+        Span span = Span.fromContext(context);
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java:100]: I know this is my code but let's extract this to a function

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java [hunk 5]
```diff
@@ -72,22 +84,61 @@ public void init(FilterConfig filterConfig) {}
...
+        Object internalRequest = extractAttr(request, "request");
+        if (internalRequest != null) {
+          byte[] postData = (byte[]) extractAttr(internalRequest, "postData");
+          if (postData != null) {
+            String post = new String(rtrim(postData), UTF_8);
+            span.setAttribute("http.request.body", post);
+          }
+        }
+        span.setAttribute("http.response.body", new String(wrapper.getContentAsByteArray(), "utf-8"));
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java]: Use UTF_8 like in line 104

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java [hunk 6]
```diff
@@ -200,4 +251,149 @@ private static String getRequestPathAttribute() {
       throw new IllegalStateException(throwable);
     }
   }
+
+  public class ContentCachingResponseWrapper extends HttpServletResponseWrapper {
```
Comment (nozik) [inline instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java:257]: this class can be private, right?

File: instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java (hunk 3 above)
Comment (LironKS) [reply instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java]: which one? the httpservletResponse ? They are doing it in many more other places

Comment (nozik) [reply instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java]: ok

Comment (LironKS) [reply instrumentation/spring/spring-webmvc-3.1/javaagent/src/main/java/org/springframework/web/servlet/OpenTelemetryHandlerMappingFilter.java]: I moved it after the validation, you'll see in a bit

**Review by LironKS — APPROVED** [review]

**Pull request merged**